src/python_be.cc

```diff
@@ -735,57 +717,36 @@ ModelInstanceState::ExecuteBLSRequest(
 
       if (pb_exception.what() != nullptr) {
         std::shared_ptr<InferPayload> infer_payload =
-            std::make_shared<InferPayload>(is_decoupled);
+            std::make_shared<InferPayload>(
+                is_decoupled, bls_response_mutex_, bls_response_cv_,
+                bls_response_buffer_);
         auto response_future =
             request_executor->Infer(infer_request, infer_payload);
-        GetBLSResponses(infer_responses, std::move(response_future));
+        infer_response = response_future.get();
 
-        size_t response_length = infer_responses.size();
-        // It is possible that the last response from the decoupled model is an
-        // empty response.
-        if (infer_responses.back() == nullptr) {
-          response_length--;
-        }
-
-        if (response_length != 1) {
-          // Construct the response_batch_shm based on the length of the
-          // responses for decoupled support.
-          response_batch_shm = Stub()->ShmPool()->Construct<char>(
-              sizeof(ResponseBatch) +
-              response_length * sizeof(bi::managed_external_buffer::handle_t));
-          response_batch =
-              reinterpret_cast<ResponseBatch*>(response_batch_shm.data_.get());
-          response_handle =
-              reinterpret_cast<bi::managed_external_buffer::handle_t*>(
-                  response_batch_shm.data_.get() + sizeof(ResponseBatch));
-          bls_response->Args() = response_batch_shm.handle_;
-          response_batch->batch_size = 1;
-          response_batch->has_error = false;
-          response_batch->is_error_set = false;
-          response_batch->cleanup = false;
-          response_batch->response_size = response_length;
-        }
+        // Need to manage the lifetime of InferPayload and RequestExecutor
+        // objects for bls decoupled responses.
+        infer_payload_[reinterpret_cast<void*>(&infer_payload)] = infer_payload;
+        request_executor_[reinterpret_cast<void*>(&infer_payload)] =
```

Comment:
Why do you need to keep the request executor alive?

Reply:
I think we would need to make sure the `TRITONSERVER_ResponseAllocator` object inside request executor is alive when it allocates for decoupled responses.

Reply:
The request executor is now stored as a member variable.